### PR TITLE
Eighth homework ready for check

### PR DESCRIPTION
**Pull request by nvkaip**

### CODE REVIEW

File: hw08/ArrayListInterpretation.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        for (index = 0; index < array.length; index++) {
+            if (array[index].equals(t)) {
+                removeHelper(index);
+                break;
+            }
+        }
+        return t;
+    }
+
+    private void removeHelper(int index) {
```
Comment (Sofasmile) [inline hw08/ArrayListInterpretation.java]: this one line method is not super useful

File: hw08/ArrayListInterpretation.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        System.arraycopy(array, index + 1, array, index, array.length - index - 1);
+    }
+
+    @Override
+    public int size() {
```
Comment (Sofasmile) [inline hw08/ArrayListInterpretation.java]: what is the size of empty list?

File: hw08/ArrayListInterpretation.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
+package com.homework.hw08;
+
+import java.util.Arrays;
+
+public class ArrayListInterpretation<T> implements List<T> {
+
+    private static final int START_SIZE = 10;
+    private Object[] array;
```
[the remark this answers is not in the document]
Comment (nvkaip) [reply hw08/ArrayListInterpretation.java]: At the beginning it was to many casts vice verse. But after I done some improvements, it's really a better idea. Just didn't notice. ))

File: hw08/ArrayListInterpretation.java (hunk 1 above)
Comment (nvkaip) [reply hw08/ArrayListInterpretation.java]: Yes, but it's prevents of code repeating. Just tried to follow this convention. Is it better to duplicate it in remove methods and why?

File: hw08/ArrayListInterpretation.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        for (++tempArrayElementIndex; tempArrayElementIndex < tempArray.length; tempArrayElementIndex++) {
+            tempArray[tempArrayElementIndex] = list.get(listElementIndex);
+            listElementIndex++;
+        }
+        array = tempArray;
+    }
+
+    @Override
+    public T get(int index) {
+        if (index >= array.length) {
```
[the remark this answers is not in the document]
Comment (nvkaip) [reply hw08/ArrayListInterpretation.java]: Exception will be thrown, isn't it?

File: hw08/ArrayListInterpretation.java (hunk 2 above)
Comment (nvkaip) [reply hw08/ArrayListInterpretation.java]: has to be 0? (with some code corrections above)

File: hw08/ArrayListInterpretation.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    private static final int START_SIZE = 10;
+    private Object[] array;
+
+    public ArrayListInterpretation() {
+        this.array = new Object[START_SIZE];
+    }
+
+    @Override
+    public void add(T value) {
```
[the remark this answers is not in the document]
Comment (nvkaip) [reply hw08/ArrayListInterpretation.java]: resolved it other way, but code became simpler

File: hw08/ArrayListInterpretation.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        int elementIndex = array.length - 1;
+        while (array[elementIndex] == null && elementIndex > 0) {
+            elementIndex--;
```
[the remark this answers is not in the document]
Comment (nvkaip) [reply hw08/ArrayListInterpretation.java]: no need of this any more

File: hw08/ArrayListInterpretation.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        if (index >= array.length - 1) {
+            array = Arrays.copyOf(array, array.length + (array.length >> 1));
+        }
+        if (array.length - 1 >= 0)
+            System.arraycopy(array, index, array, index + 1, array.length - index - 1);
+        array[index] = value;
+    }
+
+    @Override
+    public void addAll(List<T> list) {
```
[the remark this answers is not in the document]
Comment (nvkaip) [reply hw08/ArrayListInterpretation.java]: done with add(value, index)

File: hw06/FuelType.java [hunk 8]
```diff
@@ -1,6 +1,6 @@
 package com.homework.hw06;
 
-public enum FuelType{
+public enum FuelType {
     PATROL,
     DIESEL,
     LPG
```
Comment (nvkaip) [inline hw06/FuelType.java]: Do not pay attention on this, please ))

File: hw08/ArrayList.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    private static final int START_SIZE = 0;
+    private T[] array;
+
+    public ArrayList() {
+        this.array = (T[]) new Object[START_SIZE];
+    }
+
+    @Override
+    public void add(T value) {
+        array = Arrays.copyOf(array, array.length + 1);
```
Comment (Sofasmile) [inline hw08/ArrayList.java]: by default, size of the array should increase in 1.5

if you create it, you will need to use:
private int position

File: hw08/ArrayList.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+            array = Arrays.copyOf(array, index + 1);
+        }
+        System.arraycopy(array, index, array, index + 1, array.length - index - 1);
+        array[index] = value;
+    }
+
+    @Override
+    public void addAll(List<T> list) {
+        T[] tempArray = (T[]) new Object[list.size()];
+        for (int i = 0; i < (list.size()); i++) {
```
Comment (Sofasmile) [inline hw08/ArrayList.java]: do we need two for loop here?

File: hw08/ArrayList.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        return array.length < 1;
+    }
+
+    private void removeHelper(int index) {
+        System.arraycopy(array, index + 1, array, index, array.length - index - 1);
+        array = Arrays.copyOf(array, array.length - 1);
+    }
+
+    private boolean indexIsInBounds(int index) {
+        if (index >= array.length) {
```
Comment (Sofasmile) [inline hw08/ArrayList.java]: better add checking:
index < 0

File: hw08/ArrayList.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
+package com.homework.hw08;
+
+import java.util.Arrays;
+
+public class ArrayList<T> implements List<T> {
+
+    private static final int START_SIZE = 0;
```
Comment (Sofasmile) [inline hw08/ArrayList.java]: by default, start_size = 10; 
https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html

File: hw08/List.java [hunk 13]
```diff
@@ -0,0 +1,15 @@
+package com.homework.hw08;
+
+public interface List<T> {
+
+    void add(T value);
+    void add(T value, int index);
+    void addAll(List<T> list);
+    T get(int index);
+    void set(T value, int index);
+    T remove(int index); // возвращаем элемент, который удалили
```
Comment (Sofasmile) [inline hw08/List.java]: write in English please or better remove)

Comment (nvkaip) [reply hw08/List.java]: This was in German's code. I just took it, but OK, I'll clean it up
https://pastebin.com/wMsifZSt

Comment (nvkaip) [reply hw08/List.java]: This was in German's code :D. I just took it, but OK, I'll clean it up
https://pastebin.com/wMsifZSt

**Pull request merged**